src/cdsetool/download.py

```diff
@@ -28,37 +36,100 @@ def download_feature(feature, path, options=None):
+        session = _set_partial_download(session=session,size=size_temp) #here sets the header for partial downloads
         url = _follow_redirect(url, session)
         response = _retry_backoff(url, session)
-
         content_length = int(response.headers["Content-Length"])
 
         status.set_filesize(content_length)
-
-        fd, tmp = tempfile.mkstemp()  # pylint: disable=invalid-name
-        with open(tmp, "wb") as file:
+        with open(temp_path, "wb") as file:
             for chunk in response.iter_content(chunk_size=1024 * 1024 * 5):
-                if not chunk:
-                    continue
-
-                file.write(chunk)
-                status.add_progress(len(chunk))
-
-        os.close(fd)
-        shutil.move(tmp, result_path)
-
+                if chunk:
+                    file.write(chunk)
+                    status.add_progress(len(chunk))
+                
+    validity_check(Path(result_path),temp_path,feature,options,try_number)
     return feature.get("id")
 
+def validity_check(path: Path,temp_path: Path,product_info, verify_checksum: bool = True,options={},try_number = 0):
+    size = temp_path.stat().st_size
+    content_length = product_info.get("properties").get("services").get("download").get("size")
+    if size > content_length:
+            """Dowloaded more than the size of the file"""
+            temp_path.unlink()
+            raise(f"Content length mismatch for {path.name}")
+    elif size == content_length:
+        # Check integrity with MD5 checksum
+        if verify_checksum:
+            if _checksum_compare(temp_path, product_info):
+                shutil.move(temp_path, path)
+            else:
+                temp_path.unlink()
+                raise(f"Checksum mismatch for file {path.name}")
+    else:
+        """Partial dowload call back download feature"""
+        download_feature(product_info,path,options,try_number+1)
+
+def _checksum_compare(temp_path, product_info,block_size=2**13):
+    """Compare a given MD5 checksum with one calculated from a file."""
+    checksum = None
+    algo = None
+    filename = product_info.get("properties").get("title")
+    if "Checksum" in product_info:
+        checksum_list = product_info.get("Checksum")
+        if len(checksum_list) >0:
+            for checksum_dict in checksum_list:
+                if "Algorithm" in checksum_dict:
+                    algo = checksum_dict['Algorithm']
+                    if algo == "sha3-256":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.sha3_256() 
+                    elif algo == "MD5":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.md5()
+                        break
+                    elif algo == "BLAKE3":
+                        checksum = checksum_dict['Value']
+                        algo = hashlib.blake2b() # TODO ?
```

Comment:
There is no reason to loop over each algorithm. We just need the first
```python3
checksums = product_info.get("Checksum", [])
if len(checksums) == 0:
  return True # no checksum available, so we skip verifying

algo, checksum = checksums[0].get("Algorithm"), checksums[0].get("Value")

algo = {
    "sha3-256": hashlib.sha3_256(),
    "MD5": hashlib.md5(),
    "BLAKE3": blake3() # https://pypi.org/project/blake3/
}[algo]

Reply:
I really like your implementations it's clean here it's the change for the _checksum_compare()
```Python
def _checksum_compare(temp_path, product_info):
    """Compare a given MD5 checksum with one calculated from a file."""
    checksum = None
    algo = None
    filename = product_info.get("properties").get("title")
    if "Checksum" in product_info:
        # The intial query was made with the checksum option.
        checksum_list =product_info.get("Checksum")
        algo, checksum =from_checksum_list_to_checksum(checksum_list)
    else:
        odata_response = get_odata_from_product_title(filename)
        if 'value' in odata_response:
            if len(odata_response['value']) ==1:
                # odata['value'] is a list if there is more then one element for a single product
                # there is a error on the odata I guess 
                if 'Checksum' in odata_response['value'][0]:
                    checksum_list = odata_response['value'][0]['Checksum']
                    algo, checksum =from_checksum_list_to_checksum(checksum_list)
    if checksum == None:
        return None # no checksum available, so we skip verifying
    file_path = temp_path
    with open(file_path, "rb") as f:
        while True:
            block_data = f.read(8192)
            if not block_data:
                break
            algo.update(block_data)
            # status.add_progress(block_data)
    return algo.hexdigest().lower() == checksum.lower() # checksum available, so we give the verify checksum

def from_checksum_list_to_checksum(checksum_list):
    if len(checksum_list) == 0:
        return None, None # no checksum available, so we skip verifying
    algo, checksum = checksum_list[0].get("Algorithm"), checksum_list[0].get("Value")
    algo = {
        "sha3-256": hashlib.sha3_256(),
        "MD5": hashlib.md5(),
        "BLAKE3": blake3() # https://pypi.org/project/blake3/
    }[algo]
    return algo,checksum
```
Also I have added inside the pyproject.toml the following part, in which I add blake for the blake3 I don't know if it should go also somewhere else
```
dependencies = [
  "typer ~= 0.9.0",
  "rich >= 13.6,< 13.8",
  "requests >= 2.28.1,< 2.32.0",
  "pyjwt[crypto] ~= 2.8.0",
  "blake >=0.4.1" #here new depencency
]
```